aten/src/ATen/native/layer_norm.cpp

```diff
@@ -295,7 +295,13 @@ Tensor rms_norm(
       eps_val = eps.value();
     }
 
-    auto result = input.mul(at::rsqrt(at::pow(input, 2).mean(dims_to_reduce_ref, /*keep_dim=*/true).add_(eps_val)));
+    at::Tensor result;
+
+    if constexpr (std::is_same_v<scalar_t, c10::Half>) {
+      result = input.mul(at::rsqrt(at::pow(input.to(at::ScalarType::Float), 2).mean(dims_to_reduce_ref, /*keep_dim=*/true).add_(eps_val))).to(at::ScalarType::Half);
```

Comment:
@eqy what are your thoughts here?

Reply:
I also think that for bf16, we need to do some stuff in fp32.

Reply:
@mayank31398 Rsqrt on fp16 can lose a lot of precision, I'd recommend to keep it on fp32.

Reply:
> I also think that for bf16, we need to do some stuff in fp32.

BF16 has same dynamic range as fp32, but different precision. It may require separate but different handling.